tests/local/gc/gc-subtypes.wat

```diff
@@ -0,0 +1,7 @@
+;; --enable-gc
+
+(module
+  (type $a (func))
+  (type $b (sub $a (func)))
+  (type (sub final $a (func)))
+)
```

Comment:
I think we need more tests for subtyping, basically every branch of all the `Matches` implementations should be covered by our tests:

* [ ] val types that are and are not the same
* [ ] ref types and each case of same/different nullability
* [ ] heap types that are the same, not the same but sub types, and not subtypes
* [ ] each of the "known"/abstract heap types (`struct`, `noextern`, `eq`, etc)
* [ ] basically, each case in `Matches for HeapType`
* [ ] final vs not final sub types in all combinations
* [ ] super types that the sub type does vs doesn't match
* [ ] too many super types

there are a couple of these partially covered in the `gc-subtypes-invalid.wast` test, but we need to be very thorough and deliberate around exercising all these corners and combinations.

Reply:
Added more tests.